### DISCUSSION TITLE
Set colours via CSS variables in JavaScript instead of in tailwind.config.js?

### DISCUSSION THREAD

**Discussion by jameschetwood** [saadeghi/daisyui#2314] [Q&A] [answered]
If I create mytheme in tailwind.config.js I can see my buttons look great: 
```
  daisyui: {
    themes: [
      "light",
      "dark",
      {
        mytheme: {
           primary: "#f2c44f", // hsl(43, 86%, 63%)
           secondary: "#0d3fe2", // hsl(226, 89%, 47%)
           accent: "#d334b9", // hsl(310, 64%, 52%)
           neutral: "#1e191f", // hsl(290, 11%, 11%)
           "base-100": "#4b3e51", // hsl(281, 13%, 28%)
        },
      },
    ],
  },
```
<img width="652" alt="Screenshot 2023-09-01 at 09 52 48" src="https://github.com/saadeghi/daisyui/assets/6142418/cb933473-823c-421e-85c3-eb52e156434d">

However if I set different colours in tailwind.config.js and then change these later with JavaScript it doesn't fully work:
```
        mytheme: {
            primary: "pink",
            secondary: "blue",
            accent: "pink",
            neutral: "grey",
            "base-100": "#f9f9f9",
        },
```
```
    function theme() {
        const element = document.querySelector(`[data-theme= mytheme]`) as HTMLElement;
        if (!element) return;
        const b1_l = 28;
        element.style.setProperty("--p", "43 86% 63%");
        element.style.setProperty("--s", "226 89%, 47%");
        element.style.setProperty("--a", "310 64% 52%");
        element.style.setProperty("--a", "310 64% 52%");
        element.style.setProperty("--n", "290 11% 11%");
        element.style.setProperty("--b1", `281 13% ${b1_l}%`);
        element.style.setProperty("--b2", `281 13% ${b1_l - 7}%`);
        element.style.setProperty("--b3", `281 13% ${b1_l - 7 * 2}%`);
    }
```
<img width="744" alt="image" src="https://github.com/saadeghi/daisyui/assets/6142418/645817a2-59ed-4660-b32f-0a890ef345c4">

So I'm assuming that the daisyui tailwind plugin works out of the text should be light or dark? Is it possible to set a theme with just JavaScript and CSS variables in the way that I'm trying to?

**Comment by saadeghi** [comment, marked as answer]
As you can see in this table, some color names are optional:  
https://daisyui.com/docs/colors/

If those optional colors are not defined in config, they will be generated automatically based on other colors. For example, `primary-content` (`--pc`) will be generated based on `primary` (`--p`).  

If you're changing the theme in runtime, you should also define the optional colors.  
You can find their CSS variable name in this table: https://daisyui.com/docs/colors/

Let me know if you have any questions.

**Reply by jameschetwood** [reply]
Sorry I didn't give you the full context in my original question. Rather than just setting a specific theme I want to expose more controls to the user (as https://daisyui.com/theme-generator does). 

So ideally when the use sets the required colours the additional colours (eg content colours) would be auto generated, and as with normal use the system would be smart enough to know if the text should be light or dark.

To achieve this would I need to load the JS responsible from node_modules?

**Reply by saadeghi** [reply]
If you want to give something like daisyUI website's theme generator to the users, where users only pick the required color values and you want the optional colors to be generated automatically (just like daisyUI website's theme generator), you need to calculate the value for all optional colors.  
For example when user picks red color for `primary` you need to set a darker shade of red for `primary-focus` and you need to set a readable text color (black or white) for `primary-content`

Or, force the user to pick all the optional colors as well.

In daisyUI website, I used `colord` package for color manipulation and here you can find all the functions used to generate the optional colors automatically:  
https://github.com/saadeghi/daisyui/blob/master/src/docs/src/routes/theme-generator/%2Bpage.svelte


Let me know if you have any questions
